### ISSUE TITLE
openfl.net.FileReference Fails to Download when AIR.FileReference Can

### ISSUE THREAD

**Issue by rat-moonshine** [openfl/openfl#2644]
**Describe the bug**

I was try downloading a file from server, the URL looks something like this (Note: the URL returns a file to download):
https://somedomain.com/AttachmentDownloader?action=partition&id=xxxx&type=server

While downloading, each time it turns into its `IOErrorEvent` with message `Failure when receiving data from the peer.`.

While I was testing the `AIR.FileReference` codes (code-wise identical), it was able to download the expected file, though.

**To Reproduce**

Unfortunately, I don't have a public URL that I can share for the test. But the codes were mostly these:

```
private function onDownloadRequest(event:TriggerEvent):Void
{
        this.fileRef = new FileReference();
        this.attachFileReferenceListeners(true);
        this.fileRef.download(new URLRequest(this.stateData.itemURL), this.stateData.name);
}

private function attachFileReferenceListeners(value:Bool):Void 
{
        if (value) 
        {
            this.fileRef.addEventListener(Event.CANCEL, onFileBrowseCancel);
            this.fileRef.addEventListener(ProgressEvent.PROGRESS, downloadFileProgressHandler);
            this.fileRef.addEventListener(Event.COMPLETE, downloadFileCompleteHandler);
            this.fileRef.addEventListener(IOErrorEvent.IO_ERROR, downloadFileErrorHandler );
        } 
        else 
        {    
            this.fileRef.removeEventListener(Event.CANCEL, onFileBrowseCancel);
            this.fileRef.removeEventListener(ProgressEvent.PROGRESS, downloadFileProgressHandler);
            this.fileRef.removeEventListener(Event.COMPLETE, downloadFileCompleteHandler);
            this.fileRef.removeEventListener(IOErrorEvent.IO_ERROR, downloadFileErrorHandler );
        }
}

private function onFileBrowseCancel(event:Event):Void
{
}

private function downloadFileProgressHandler(event:ProgressEvent):Void
{
        this.hProgress.maximum = event.bytesTotal;
        this.hProgress.value = event.bytesLoaded;
}

private function downloadFileCompleteHandler(event:Event):Void
{
        this.attachFileReferenceListeners(false);
}

private function downloadFileErrorHandler(event:IOErrorEvent):Void
{
        this.attachFileReferenceListeners(false);
}
```

**Expected behavior**

File should download without raising an error.

**OpenFL Targets**

CPP, macOS

**Additional context**

Testing on:
- Openfl 9.3.0-Dev
- Lime 8.1.0-Dev 
- macOS Ventura

**Comment by joshtynjala** [comment]
> While downloading, each time it turns into its IOErrorEvent with message Failure when receiving data from the peer..

Unfortunately, I haven't been able to reproduce this issue. I tried downloading a variety of text, image, and other binary files, but I could download them all without that error message being reported. (In the process, I found a separate issue with binary files that I fixed in commit 23e79eee7597d40a4d0957b5dfa010c89ee70fc3, but I don't believe that is related.)

From some quick research, that error message comes from the curl library. Some folks said that it could indicate a backend issue, but if the download works with AIR, it seems to indicate an issue in Lime/OpenFL.

I can't do much more until we can figure out how to get it to reproduce on my computer. @rat-moonshine You are welcome to email me a URL privately, if that's possible.

**Comment by rat-moonshine** [comment]
Thank you @joshtynjala . I'm discussing here if we can arrange you a similar download-URL that do not requires you to authenticate securely.

**Comment by JoelProminic** [comment]
@joshtynjala, we can't provide the original download URL, but I can give some more information.  Note that the file is not downloaded directly (the filename is not in the URL), but it is indirectly downloaded through a servlet for security.  The headers look like this:

```
response.setContentType("application/octet-stream");

response.setHeader("Pragma", "cache");
response.setHeader("Cache-control", "public");
response.setHeader("Expires:", formatter.format( expire_cal.getTime()));
response.setHeader("Content-Disposition", "attachment; filename=" + attachmentName);
response.setHeader("Content-Length", attachmentLength);
```

This code is a bit old, but it has been working fine from my recent tests in web browsers.

If necessary, I should be able to create a test action that returns a test file with these headers.

**Comment by joshtynjala** [comment]
I don't do a lot of backend work, so someone creating a test URL would be ideal for me.

**Comment by JoelProminic** [comment]
Sorry for the delay @joshtynjala, here is an example agent.  In this case it is returning a text file instead of binary data, but I can convert it to a binary file if needed.

https://rest-swf.prominic.net/Flex.nsf/TestDownload?OpenAgent

The expected content of the downloaded file is "Hello, world!"

**Comment by joshtynjala** [comment]
Thanks, @JoelProminic. I'll try to take a look soon.

**Comment by joshtynjala** [comment]
I can successfully use `FileReference` to download from the example agent URL. It does not seem to reproduce the issue. I used the same code as above, except to change the URL:

```haxe
this.fileRef.download(new URLRequest("https://rest-swf.prominic.net/Flex.nsf/TestDownload?OpenAgent"), "file.txt");
```

If the original file is binary, perhaps that would make a difference.

**Comment by JoelProminic** [comment]
I had a little trouble making a good binary example, but you can try this:
https://rest-swf.prominic.net/Flex.nsf/TestDownloadBinary?OpenAgent

@rat-moonshine, if @joshtynjala can't reproduce this with my URLs, please try a quick test with the same URLs on your side (i.e. hard code the download URL temporarily).

**Comment by rat-moonshine** [comment]
To my surprise, when I tested today I found the secured link downloads correctly when using with `FileReference`. But I remember I had trouble earlier downloading from the said URL. 

I don't know what caused this change, or why I had the earlier error. I did updates Openfl, Lime couple of time by then, I don't know if that makes any difference. 

For now, I can download correctly from the type of URL. Perhaps we can close this now, until we encounter this again. Sorry, this turned unexpectedly.

**Comment by joshtynjala** [comment]
> I don't know what caused this change, or why I had the earlier error. I did updates Openfl, Lime couple of time by then, I don't know if that makes any difference.

If it was a binary file, then perhaps 23e79eee7597d40a4d0957b5dfa010c89ee70fc3 was the required fix afterall.

**Comment by rat-moonshine** [comment]
I starts seeing the download problem with `OpenFL.FileReference` again when `AIR.FileReference` downloads it correctly. 

I see following error-text with `IOErrorEvent.IO_ERROR`:

![image](https://github.com/openfl/openfl/assets/24370870/dc35b63c-fea1-49a3-a2a4-9ba8ea71aa27)

I'm still using same code as mentioned int he [original](https://github.com/openfl/openfl/issues/2644#issue-1737472380) issue. Do I need to add anything additional, to cover such scenario @joshtynjala (?)

**Comment by joshtynjala** [comment]
@rat-moonshine I cannot reproduce by using the original code with the following test URL shared above: _https://rest-swf.prominic.net/Flex.nsf/TestDownloadBinary?OpenAgent_

OpenFL 9.3.0, Lime 8.1.0, Haxe 4.3.2

I get a couple of progress events and a complete event. No IO error events. The contents of the downloaded file from the mac target matches the air target.

**Comment by rat-moonshine** [comment]
We talked on this and we found that there was a download restriction - to the URL that I was trying to download from. 

We also noticed that the agent returns an XML with related information in that case.

I tried to catch that XML using `DataEvent.DATA` (or, even by `DataEvent.UPLOAD_COMPLETE_DATA`), unfortunately the usage didn't returned anything but falls into `IOErrorEvent.IO_ERROR`.

The only message console prints in that case was this:

`2023-10-27 11:11:31.038 osascript[7772:87100] +[CATransaction synchronize] called within transaction`

Is there any way we can track the returning XML/data @joshtynjala (?)

**Comment by joshtynjala** [comment]
I'll check if Flash/AIR returns that data somehow. If it does, I'll ensure that OpenFL does it too.

**Comment by rat-moonshine** [comment]
Thank you @joshtynjala , but we found something more interesting.

With `AIR.FileReference`, the API somehow generates/saves the file on user system, however when opened that file in a text-editor I saw this:

![image](https://github.com/openfl/openfl/assets/24370870/1fa01915-9b7a-4508-a2f3-3c4b82cc9c8a)

This is the same XML that we was expecting to return as a `DataEvent`, but here beside getting into an IO/Error-event AIR saves the XML in the generated/saved file only.

In case of `OpenFL.FileReference` however it turns into IO/Error directly.
